### DISCUSSION TITLE
How to augment `useNuxtApp` type?

### DISCUSSION THREAD

**Discussion by Barbapapazes** [nuxt/nuxt#25590] [Questions]
Hello,

I'm doing a module where I use a plugin to inject data in `useNuxtApp`.

In another plugin, I do `const { $plausible } = useNuxtApp()`. How to type (if possible) $plausible?

**Comment by arashsheyda** [comment]
@Barbapapazes meybe something like:

```

declare module '#app' {
  interface NuxtApp {
    $plausible: typeof x
 }
}

```

**Reply by adamdehaven** [reply]
This doesn't seem to work, at least not within a project utilizing Nuxt layers

**Reply by dsvgl** [reply]
The `ComponentCustomProperties` part makes it so that `$plausible` is declared when used in template part, too.

```ts
// myModule/src/types/template.d.ts

interface Plausible {
  $plausible: typeof x;
}

declare module '#app' {
  interface NuxtApp extends Plausible {}
}

declare module 'vue' {
  interface ComponentCustomProperties extends Plausible {}
}

// webstorm issue: https://youtrack.jetbrains.com/issue/WEB-59818/Vue-custom-global-properties-added-by-augmenting-vue-are-not-resolved
declare module '@vue/runtime-core' {
  interface ComponentCustomProperties extends Plausible {}
}

export {};
```

```ts
// myModule/src/module.ts

import { addTypeTemplate, defineNuxtModule } from '@nuxt/kit';

export default defineNuxtModule<ModuleOptions>({
  // ...
  setup(moduleOptions, nuxt) {
    addTypeTemplate({
      filename: 'my-module-augmentations.d.ts',
      src: resolve('./types/template.d.ts'),
    });
  },
});
```

**Reply by adamdehaven** [reply]
The only way the type resolves for me is to change `#app` to `nuxt/app`

```ts
declare module 'nuxt/app' {
```

**Reply by dsvgl** [reply]
> The only way the type resolves for me is to change #app to nuxt/app

In our monorepo (multiple nuxt apps and nuxt layers) it works fine with `#app`. Are you on the latest Nuxt version (3.11.2 atm)?

**Reply by adamdehaven** [reply]
Odd, I got it working this morning with the [recommended way](https://nuxt.com/docs/guide/directory-structure/plugins#typing-plugins) 🚀 

Not sure what I had going on previously

**Reply by adamdehaven** [reply]
Ok I take this back; it's not actually working for me. 

I didn't initially realize you wrote a separate module to actually inject the types. I believe this _does_ work; however, if I do _not_ inject via a module, and simply add the file shown below, my project does not correctly pick up the interface without _also_ supplying a module for `nuxt/app`

```typescript
interface NuxtAppHelpers {
  $api: $Fetch
  $fetchWithCookie: (event: H3Event, url: string, options: any = {}) => Promise<Response>
}

// This does not work in my project
declare module '#app' {
  interface NuxtApp extends NuxtAppHelpers {}
}

// This **does** work and the types correctly resolve
declare module 'nuxt/app' {
  interface NuxtApp extends NuxtAppHelpers {}
}
```

**Reply by dsvgl** [reply]
Ah, good to know.
